compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -418,11 +419,47 @@ impl<'hir> LoweringContext<'_, 'hir> {
         let header = self.generate_header_error();
         let sig = hir::FnSig { decl, header, span };
 
-        let ident = Ident::dummy();
-        let body_id = self.lower_body(|this| (&[], this.mk_expr(hir::ExprKind::Err(err), span)));
+        let ident = self.lower_ident(delegation.ident);
+
+        let body_id = self.lower_body(|this| {
+            let body_expr = match delegation.body.as_ref() {
+                Some(box block) => {
+                    hir::ExprKind::Block(this.generate_delegation_error_block(block, span), None)
+                }
+                None => hir::ExprKind::Err(err),
+            };
+
+            (&[], this.mk_expr(body_expr, span))
+        });
+
         DelegationResults { ident, generics, body_id, sig }
     }
 
+    // Generates a block when we failed to resolve delegation, where a target expression is its only statement,
+    // thus there will be no ICEs on further stages of analysis (see #144594)
+    fn generate_delegation_error_block(
```

Comment:
Nit: the function is small enough and only used once, it can be inlined and removed.